File: 단계별로 풀어보기/Step 16/2156.cpp
#include <iostream>
using namespace std;

int main() {
    int n;
    int wine[10003];
    int dp[10003];
    int i;

    cin >> n;
    for(i=3; i<=n+2; i++) {
        cin >> wine[i];
    }

    for(i=3; i<=n+2; i++) {
        dp[i] = max(dp[i-1], wine[i]+dp[i-2]);
        dp[i] = max(dp[i], wine[i]+wine[i-1]+dp[i-3]);
    }

    cout << dp[n+2] << endl;
    return 0;
}

File: StepByStep/Python/Step_16_동적 계획법 1/2156.py
import sys

N = int(input())
wine = [0] * 10000
dp = [0] * 10000
for i in range(N):
    wine[i] = int(sys.stdin.readline())

dp[0] = wine[0]
dp[1] = wine[0] + wine[1]
dp[2] = max(wine[0] + wine[1], wine[0] + wine[2], wine[1] + wine[2])
for i in range(3, N):
    dp[i] = max(dp[i-3]+wine[i-1]+wine[i], dp[i-2]+wine[i], dp[i-1])

print(dp[N-1])
